src/main/java/org/cryptacular/generator/AESP12Generator.java

```diff
@@ -0,0 +1,200 @@
+    DIGEST_ID_TO_HMAC_ID_MAP.put(NISTObjectIdentifiers.id_sha3_512, NISTObjectIdentifiers.id_hmacWithSHA3_512);
+  }
+
+
+  /**
+   * Creates a new instance that encrypts with AES-256-CBC and SHA256 using 2048 rounds of hashing.
+   */
+  public AESP12Generator()
+  {
+    this(NISTObjectIdentifiers.id_sha256, 2048);
+  }
+
+  /**
+   * Creates a new instance that encrypts with AES-256-CBC and SHA256 using the given number of hashing rounds.
+   *
+   * @param iterations Number of rounds of encryption.
+   */
+  public AESP12Generator(final int iterations)
+  {
+    this(NISTObjectIdentifiers.id_sha256, iterations);
+  }
+
+  /**
+   * Creates a new instances that uses AES-256-CBC and the given digest algorithm to encrypt data.
+   *
+   * @param digestAlgId Digest algorithm identifier.
+   * @param iterations Number of rounds of hashing.
+   */
+  public AESP12Generator(final ASN1ObjectIdentifier digestAlgId, final int iterations)
+  {
+    if (!SUPPORTED_DIGEST_ALGORITHMS.contains(digestAlgId)) {
+      throw new IllegalArgumentException("Unsupported digest algorithm");
+    }
+    digestAlgorithm = digestAlgId;
+    final ASN1ObjectIdentifier hmacAlgId = DIGEST_ID_TO_HMAC_ID_MAP.get(digestAlgId);
+    // The default behavior of the builder is to select salt size based on HMAC algorithm,
+    // which is the desirable behavior here
+    pbkdf2Config = new PBKDF2Config.Builder()
+      .withIterationCount(iterations)
+      .withPRF(new AlgorithmIdentifier(hmacAlgId, DERNull.INSTANCE))
+      .build();
+    outputEncryptorBuilder = new PBES2OutputEncryptorBuilder(PBES2Algorithm.AES256, pbkdf2Config);
+  }
+
+  @Override
+  public int getIterations()
+  {
+    return pbkdf2Config.getIterationCount();
+  }
+
+  @Override
+  protected ASN1ObjectIdentifier getDigestAlgorithmId()
+  {
+    return digestAlgorithm;
+  }
+
+  @Override
+  protected DigestSpec getDigestSpec()
+  {
+    return DIGEST_ID_TO_DIGEST_SPEC_MAP.get(digestAlgorithm);
+  }
+
+  @Override
+  protected OutputEncryptor keyOutputEncryptor(final char[] password)
+  {
+    return outputEncryptorBuilder.build(password);
+  }
+
+  @Override
+  protected OutputEncryptor dataOutputEncryptor(final char[] password)
+  {
+    return outputEncryptorBuilder.build(password);
+  }
+
+  /**
+   * Builds an output encryptor based on PBKDF2 domain parameters.
+   */
+  static class PBES2OutputEncryptorBuilder
+  {
+    /** Source of cryptographically-strong randomness. */
+    private final SecureRandom random = new SecureRandom();
+
+    /** PBES2 encryption algorithm. */
+    private final PBES2Algorithm encryptionAlg;
+
+    /** PBKDF2 domain parameters. */
+    private final PBKDF2Config pbkdf2Config;
+
+
+    PBES2OutputEncryptorBuilder(final PBES2Algorithm encAlg, final PBKDF2Config config)
+    {
+      this.encryptionAlg = encAlg;
+      this.pbkdf2Config = config;
+    }
+
+    public OutputEncryptor build(final char[] password)
```

Comment:
Add check for null or empty password?

Reply:
Added a null/empty check in the public `generate` method shared by both classes and reduced the visibility of this method to emphasize that it's an implementation detail.